src/Elliptic/DiscontinuousGalerkin/SubdomainOperator/SubdomainOperator.hpp

```diff
@@ -0,0 +1,641 @@
+#include "ParallelAlgorithms/LinearSolver/Schwarz/ElementCenteredSubdomainData.hpp"
+#include "ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp"
+#include "ParallelAlgorithms/LinearSolver/Schwarz/SubdomainOperator.hpp"
+#include "ParallelAlgorithms/LinearSolver/Schwarz/Tags.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// Items related to the restriction of the DG operator to an element-centered
+/// subdomain
+namespace elliptic::dg::subdomain_operator {
+
+namespace detail {
+// Wrap the `Tag` in `LinearSolver::Schwarz::Tags::Overlaps`, except if it is
+// included in `TakeFromCenterTags`.
+template <typename Tag, typename Dim, typename OptionsGroup,
+          typename TakeFromCenterTags>
+struct make_overlap_tag_impl {
+  using type = tmpl::conditional_t<
+      tmpl::list_contains_v<TakeFromCenterTags, Tag>, Tag,
+      LinearSolver::Schwarz::Tags::Overlaps<Tag, Dim::value, OptionsGroup>>;
+};
+
+// Wrap the `Tag` in `Tags::NeighborMortars`
+template <typename Tag, typename Dim>
+struct make_neighbor_mortars_tag_impl {
+  using type = Tags::NeighborMortars<Tag, Dim::value>;
+};
+}  // namespace detail
+
+/*!
+ * \brief The elliptic DG operator on an element-centered subdomain
+ *
+ * This operator is a restriction of the full (linearized) DG-operator to an
+ * element-centered subdomain with a few points overlap into neighboring
+ * elements. It is a `LinearSolver::Schwarz::SubdomainOperator` to be used with
+ * the Schwarz linear solver when it solves the elliptic DG operator.
+ *
+ * This operator requires the following tags are available on overlap regions
+ * with neighboring elements:
+ *
+ * - Geometric quantities provided by
+ *   `elliptic::dg::subdomain_operator::InitializeSubdomain`.
+ * - All `System::fluxes_computer::argument_tags` and
+ *   `System::sources_computer::argument_tags` (or
+ *   `System::sources_computer_linearized::argument_tags` for nonlinear
+ *   systems), except those listed in `ArgsTagsFromCenter`. The latter will be
+ *   taken from the central element's DataBox, so they don't need to be made
+ *   available on overlaps.
+ * - The `System::fluxes_computer::argument_tags` on internal and external
+ *   interfaces, except those listed in `System::fluxes_computer::volume_tags`.
+ *
+ * Some of these tags may require communication between elements. For example,
+ * nonlinear system fields are constant background fields for the linearized DG
+ * operator, but are updated in every nonlinear solver iteration. Therefore, the
+ * updated nonlinear fields must be communicated across overlaps between
+ * nonlinear solver iterations. To perform the communication you can use
+ * `LinearSolver::Schwarz::Actions::SendOverlapFields` and
+ * `LinearSolver::Schwarz::Actions::ReceiveOverlapFields`, setting
+ * `RestrictToOverlap` to `false`.
+ *
+ * \warning The subdomain operator hasn't been tested with periodic boundary
+ * conditions so far.
```

Comment:
Yes, the subdomain operator has to deal with _orientations_ a lot: Data on overlap regions is always stored as if it was the DataBox of the overlapped neighbor, so the logical axes may not be aligned between data on central element and overlap regions. Therefore I have to re-orient the data whenever it crosses element boundaries (there's a note on that in the `LinearSolver::Schwarz::Tags::Overlaps` docs).

But perhaps you mean something other than orientations? The logical axes may have kinks at element boundaries, but that's not a problem for the subdomain operator (why is it a problem for subcell?). The tests show that the subdomain operator is equivalent to the full DG operator on a cylinder, which has kinks in the logical axes.

Reply:
Yea, it's the kinks I don't understand. In subcell it's an issue (or adds complexity) because one elements grid point at `(1.2,0,0)` (just something arbitrary but in a neighbor) is not equal to the neighbor's `(-0.8,0,0)`, it has some different 3d coordinates. This means the reconstruction and FD derivatives need to take this kink into account. I don't know exactly what the subdomain operator does, so that's why I'm asking about this. Basically, on a curved mesh we are no longer using a structured mesh, but rather a block-structured mesh, which means there are parts of the grid that need to be treated as unstructured (block boundaries). This might not matter for how well the preconditioner works, though.

Reply:
Ah I see. These kinks are not an issue for the subdomain operator because it just calls into the standard DG operator. Therefore, when the standard DG operator is OK with the kinks, so is the subdomain operator :) Thanks for raising this though!

Reply:
That's the part I don't understand. Doesn't the standard DG operator involve taking derivatives? In which case isn't it assuming some point distribution?

Reply:
ah, perhaps the confusion is this: The subdomain operator doesn't treat the central element plus overlap regions as a single "combined" mesh. It retains the element structure including faces and mortars. The only difference to the standard DG operator is that all data is local, so exchanging data across mortars requires no communication with other processors.

Reply:
ah, I (think) see. So the neighboring element data that is included in the mesh is not updated, and you can do iterations on the local data to convergence?

Reply:
yea, here's a picture that might help (from https://spectre-code.org/structLinearSolver_1_1Schwarz_1_1Schwarz.html):

![subdomain_structure](https://user-images.githubusercontent.com/746230/118534748-de007680-b749-11eb-8c80-ded06268d2c3.png)

The points within the dashed line are part of the subdomain. Every other point is zero for the subdomain operator. The subdomain solver iterates on this data until it has found a solution; this happens locally on an element (no communications). Then the solution on overlap regions is sent to the neighbors and weighted with the solutions from other subdomains to obtain an approximate solution over the entire domain.

Reply:
:+1: okay, that makes sense. Thanks for explaining that to me :)